### PR TITLE
Add apu3 to led-apu driver

### PR DESCRIPTION
**Pull request by feckert**
The APU3 is optimized for 3G/LTE modem.

- It has an additional GPIO for SIM switching.
- The board name has changed from APU2 to APU3.
- GPIO-Leds and GPIO-Buttons have not changed.

See link from vendor
http://www.pcengines.ch/apu2.htm

### CODE REVIEW

**Comment by riptidewave93** [comment]
@feckert According to the info shared at https://github.com/riptidewave93/LEDE-APU2/issues/18#issuecomment-313903620, it seems this patch my not apply to the latest APU3 BIOS. Can you confirm it's been tested on BIOS 4.5.5?

**Comment by feckert** [comment]
@riptidewave93 sorry i have only a APU3 board with version "Version: 4.0.7" and "Release Date: 03/02/2017". On this board the
Manufacturer is "PC Engines" and the Product Name is "APU3".
Is there a newer BIOS available?

**Comment by riptidewave93** [comment]
@feckert Seems there is also 4.5.5 as "experimental" as it's based on Coreboot mainline, instead of 4.0. http://pcengines.ch/howto.htm#bios

**Comment by feckert** [comment]
@riptidewave93 i think as long as the BIOS is marked as experimental the driver should only load on stable BIOS.
Independently i think PC Engine should change the "/sys/devices/virtual/dmi/id/product_name" to "APU3" and not to "PC Engine APU3" as reported by @riodda. This will also prevent the driver "kmod-gpio-nct5104d" from loading.

I could also change the check on this driver from https://github.com/lede-project/source/pull/1206/files#diff-8c650613f8344e55ef6a33f7176268daR338 to "APU" as in the "kmod-gpio-nct5104d" driver
https://github.com/lede-project/source/blob/master/package/kernel/gpio-nct5104d/src/gpio-nct5104d.c#L436.

What to you think

**Comment by riptidewave93** [comment]
@feckert I am fine with only supporting "Stable" BIOS releases.

I am curious if @pietrushnic has any input on the name change in the "experimental" release, as he works on the BIOS for the APU units.

As for moving forward, moving to just "APU" could also mean this would try to load on the APU1, which would fail. I think the current approach using strcasecmp on apu2/apu3 is the best for the time being.

**Comment by pietrushnic** [comment]
@riptidewave93 what we have in DMI (precisely in SMBIOS) is driven by coreboot standard way of exposing that information. coreboot developers will not merge changes that not follow that rules.

Mainline releases 4.5.x and coming 4.6.x are marked as experimental by PC Engines, because those firmware versions were not tested enough to be recommended. Those releases should not be used in production environment yet. Despite that I would be glad to get more feedback about mainline releases.

Legacy releases 4.0.x are built on top of old code and do not comply with coreboot standard in various areas ie. naming.

According to my knowledge manufacturer name should be `PC Engines` and product name should be `apu2`. I cannot change already released fimrware images, but I can fix what will be released in future.

**Comment by feckert** [comment]
@pietrushnic 

> According to my knowledge manufacturer name should be PC Engines and product name should be apu2. I cannot change already released fimrware images, but I can fix what will be released in future.

That means it stayes as it is in the actual bios
manufactuerer: PC Engines
product: "apu2" -> for apu2 and "apu3" ->for apu3
?

**Comment by pietrushnic** [comment]
@feckert yes. If you see something else like `PCEngines` in v4.5.5, then it is a bug.

**Comment by pietrushnic** [comment]
@feckert I cannot check sysfs now since I'm using Voyage Linux. Problem with `PCEngines` applies only to v4.5.5. Name fix was introduced for APU2 and APU3 in [v4.5.7](https://github.com/pcengines/coreboot/blob/coreboot-4.5.x/CHANGELOG.md#v457---2017-06-08). For my APU2 v4.5.8 DMI looks like that:

```
root@voyage:~# dmidecode -t system
# dmidecode 2.12
SMBIOS 2.7 present.

Handle 0x0001, DMI type 1, 27 bytes
System Information
        Manufacturer: PC Engines
        Product Name: PC Engines apu2
        Version: 1.0
        Serial Number: 1069064
        UUID: Not Settable
        Wake-up Type: Reserved
        SKU Number: 4 GB
        Family: Not Specified

Handle 0x0005, DMI type 32, 11 bytes
System Boot Information
        Status: No errors detected
```

I know dealing with this entries breaks things. If you will find inconsistency please let me know and where I can I will fix those.

File: package/kernel/leds-apu2/src/leds-apu2.c [hunk 1]
```diff
@@ -69,7 +69,7 @@ MODULE_DEVICE_TABLE (pci, gpio_apu2_pci_tbl);
 /* EGPIO89=GPIO32, EGPIO90=GPIO33, AGPIO68=GPIO57, AGPIO69=GPIO58, AGPIO70=GPIO59 */
 static u8 gpio_offset[APU_NUM_GPIO] = {89, 90, 68, 69, 70};
 
-static void __iomem *gpio_addr[APU_NUM_GPIO] = {NULL, NULL, NULL, NULL};
+static void __iomem *gpio_addr[APU_NUM_GPIO] = {NULL, NULL, NULL, NULL, NULL};
```
Comment (hauke) [inline package/kernel/leds-apu2/src/leds-apu2.c]: This should go into the previous commit

File: package/kernel/leds-apu2/src/leds-apu2.c [hunk 2]
```diff
@@ -251,7 +251,7 @@ static struct gpio_keys_button apu2_gpio_keys[] = {
 		.type           = EV_KEY,
 		.code           = KEY_RESTART,
 		.debounce_interval = 60,
-		.gpio           = 508,
+		.gpio           = 507,
```
Comment (hauke) [inline package/kernel/leds-apu2/src/leds-apu2.c]: Does this still work on the APU2?

Comment (pietrushnic) [inline package/kernel/leds-apu2/src/leds-apu2.c]: @mek-x any idea bout that ?

Comment (feckert) [inline package/kernel/leds-apu2/src/leds-apu2.c]: Sorry i have no apu2 board to test this and if this works!

File: package/kernel/leds-apu2/src/leds-apu2.c (hunk 1 above)
Comment (feckert) [inline package/kernel/leds-apu2/src/leds-apu2.c]: Will fix it

**Comment by feckert** [comment]
@pietrushnic 
Sorry again for the stupid question. Could you please clarify.

>         Manufacturer: xxx
>         Product Name: xxx

for the different BIOS Version now and in the future.

**Comment by pietrushnic** [comment]
@feckert I have email conversation with PC Engines about confirming exactly what we will have, but at this point I plan to have:

```
        Manufacturer: PC Engines
        Product Name: PC Engines apu2
```

for apu2. Of course for apu3 and following product name will change.

**Comment by pietrushnic** [comment]
@CRCinAU :) This very old firmware. We have not chance of changing old binaries that were released. I know this is a mess, but there was no standard way of representing this information. We were not aware that someone rely on this fields. If you want to support legacy systems you have to match couple combinations.

**Comment by CRCinAU** [comment]
Is there any compelling reason to upgrade from this version to newer? apart from possibly breaking the module support for LEDs / GPIO etc? :)

**Comment by pietrushnic** [comment]
@CRCinAU it depends how you use your system and what you expect. We maintain [legacy](https://github.com/pcengines/release_manifests/blob/coreboot-4.0.x/CHANGELOG.md) and [mainline](https://github.com/pcengines/release_manifests/blob/coreboot-4.5.x/CHANGELOG.md) releases and you can check CHANGELOGs. It's your judgment if some features have value to you.

**Comment by CRCinAU** [comment]
Thanks for the info. I mainly use mine as a router - so the main function is that it just boots and routes packets... As such, I guess bios upgrades aren't really critical for me - as long as it'll boot afterwards.

Congrats on working on getting mainline operational on the APU2... It's a much better goal than trying to maintain your own fork.

I introduced The Boss at work to the APU series - and we've got one running, one for an employee, and one (possibly 2) in transit - mainly due to how helpful everyone is with these bits of kit!

**Comment by pietrushnic** [comment]
@CRCinAU As I mentioned in other threads I'm always open for support of this platform. Sometimes especially in heavy usage there may be need for advanced features of this platform (eg. ECC, secure/verified boot, AMD-V, EVP), if you will face firmware limitation we are always ready to discuss that and help if possible.

**Comment by riodda** [comment]
Any news about merging this ? I confirm that the patch brings back the leds on APU3.

**Comment by feckert** [comment]
If update the pull request as suggested by @hauke .

- I have no apu2 to test if the gpio_offset will work on apu2 (LEDS). I have changed the offset because of the SIM-SWITCH gpio.

- If some one could also test the SIM-SWITCH gpio if this works as expected! On my test board i could switch the gpio but the SIM will not be detected!

- If everything is fine i think we could merge this. If in the future the manufacture or the name of the board will change we could update the board detection code.

**Comment by feckert** [comment]
@riptidewave93 any progress on this PR?

**Comment by riptidewave93** [comment]
@feckert Sorry, I have been busy with other projects. I reached out to @CRCinAU and he is going to verify the reset button still works on the APU2 after the changes added at https://github.com/lede-project/source/pull/1206/files#diff-8c650613f8344e55ef6a33f7176268daR254

**Comment by CRCinAU** [comment]
Bad news, all the LEDs disappear when using this patch on the APU2.

Using BIOS 4.0.7 as per above, and getting the following from dmidecode:
```
$ dmidecode -t system
# dmidecode 3.0
Getting SMBIOS data from sysfs.
SMBIOS 2.7 present.

Handle 0x0001, DMI type 1, 27 bytes
System Information
        Manufacturer: PC Engines
        Product Name: APU2
```

Interestingly, if I change APU_NUM_GPIO to 4, then all works. Even though we define 5 elements etc. I'll leave that one for folks to sort out....

**Comment by feckert** [comment]
@riptidewave93 @CRCinAU  sorry for the late replay i was also busy with other things!
To get on with this PR i think we should remove the SIM commit and only add this two commits
https://github.com/lede-project/source/pull/1206/commits/fd3cc896b971d5056586b8cee01b0a75ec6d0d7f
https://github.com/lede-project/source/pull/1206/commits/600cbe04085bd0eaa7d28f7d432786e5e8a7a7d4
This will fix the led on APU3 and hopfully works also for APU2.
SIM switching could be done on a next PR?

**Comment by riodda** [comment]
the problem is that if you remove the sim sw part the sim switch GPIO will stay floating and this will make impossible to use any of the mpci slot for mobile conneciton at least this is the problem i got.

**Comment by pietrushnic** [comment]
@riodda I'm not sure if this would help, but we have 2 releases with SIMSWAP (GPIO33) set to low or high by default. Please check:

https://github.com/pcengines/coreboot/blob/coreboot-4.6.x/CHANGELOG.md#v4552---2017-03-03

and 

https://github.com/pcengines/coreboot/blob/coreboot-4.6.x/CHANGELOG.md#v455---2017-02-24

@mek-x can you confirm that ?

**Comment by mek-x** [comment]
Yes, versions `4.5.5.1` and `4.5.5.2` differ only with different state of `simswap` pin.

**Comment by feckert** [comment]
I think we should split up the driver into different parts.
One into a common gpio driver part (gpio-fch) and the others into platform driver which use the gpio-fch with platform drivers.

I have pushed a RFC driver to https://github.com/TDT-GmbH/source/tree/gpio-fch
This driver should show all gpios for the "amd fch southbridge". The driver based on this is https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/drivers/gpio/gpio-amd8111.c?h=v4.4.89 in the upstream kernel.

The Problem is that i am unable to load this driver because I always get the message from my driver that this region is already used https://github.com/TDT-GmbH/source/commit/c2a294d7a3063f9705ce5d0c5e43178c529a0cfc#diff-ff120ae031a52ce88103c487a2ee0cb5R291. May be someone could help me to get this fixed to go ahead. I am not a pci guru so i do not know why this region is already used! Under "/proc/bus/pci/devices" there is no driver listed which uses the region.

Any hints or comments to go ahead with this?

**Comment by CRCinAU** [comment]
I've kinda not been following this too closely - but do we have a working setup for APU2 & APU3 yet?

I do add the APU2 patch into my kernel builds for Xen - but since we started playing, I've got a non-working patch now...

As such, I'd like to update the patch I use to support both APU2 & APU3...

https://xen.crc.id.au/git/?p=kernel-xen;a=tree;f=SOURCES;h=a16febad32fcf7c14ae239da70bf155b413290f6;hb=HEAD

**Comment by CRCinAU** [comment]
ping?

**Comment by hauke** [comment]
The first two patches were applied to master, Please improve the last patch to make it still work with APU2

**Comment by feckert** [comment]
@hauke thanks for merging
@CRCinAU I have no APU2 so i am unable to test this. Have you tested your changes with apu2 and apu3 if so ( and it is working ) i will update this PR with your changes

**Comment by CRCinAU** [comment]
@feckert My version of the driver is currently broken after attempting to merge the APU3 bits with it...

The last known working version for me is:
https://xen.crc.id.au/git/?p=kernel-xen;a=commit;h=3e78264b2a0fd21055b1eff787eda1d90768598a

Anything beyond that is broken as per my comments on Aug 24 on this thread...

**Comment by feckert** [comment]
@hauke I think we could close this PR.
If we need the sim pin then a new PR should be opened again.
Leds should work on apu2 and apu3 now.

**Comment by blogic** [comment]
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.

**Comment by feckert** [comment]
Hello @hauke @CRCinAU @riptidewave93 @pietrushnic 

I have written a new gpio-apu device driver which will uses the GPIOs found on APU2/APU3.
https://github.com/TDT-AG/openwrt/tree/pr/20180730-kernel-package-add-gpio-apu-driver
I will send this also upstream to the kernel on positive feedback.

Explanation:

- remove old leds-apu2 driver which is not upstream compatible
- backport new led-apu driver from kernel upstream https://github.com/torvalds/linux/blob/master/drivers/leds/leds-apu.c
- add dmidecode information to backported leds-apu driver for APU3
- add new gpio-apu driver for APU2 and APU3 (reset button for APU2/APU3, sim switch for APU3 only)

**Comment by pietrushnic** [comment]
@miczyg1 can we test that and provide feedback? Also if you can take a look at code it would be great.

@feckert I assume it makes [linux-gpio-nct5104d](https://github.com/pcengines/linux-gpio-nct5104d) deprecated?

**Comment by miczyg1** [comment]
@pietrushnic Yes, will be glad to do it. Maybe we should provide support for apu4 too? @feckert apu4 is pretty the same board as apu3 (apu4 has 4 network controllers). I will check the schematics for any differences in GPIOs.

**Comment by feckert** [comment]
@pietrushnic no this will not deprecated the nct5104d this is driver which is responsible for the supperIO chips. The gpio-apu driver is responsible for the gpios on the CPU it self. This are the reset button on APU2/APU3 and the SIM switch gpio on APU3. 
@pietrushnic if this has also reset button or sim switch we can also add this to the demidecode information.

**Comment by riptidewave93** [comment]
@feckert have you PR'd this against the OpenWRT repo yet? Once done I would be glad to review and offer a +1 on my side.

**Comment by feckert** [comment]
@riptidewave93 @pietrushnic @miczyg1 
I have add a new branch for the pullrequest against openwrt master.
See:
https://github.com/openwrt/openwrt/pull/1232

**Comment by miczyg1** [comment]
@feckert I have compiled OpenWrt with Your leds and gpio modules. LEDs are working allright. I can manually set all 3 leds on or turn them off via sysfs. I wonder what is this button:
https://github.com/openwrt/openwrt/pull/1232/files#diff-e91e2c91ca210d7d59244a8f04527503R191

Do You mean the S1 button on the back side of board? The code suggests that it should reset the platform. However it is not doing so. I would appreciate if You could provide information about features introduced in these patches and how to verify them.

@feckert also I suggest to add support for apu4 boards. They have the same GPIO layout as apu3. Have a look here (page 7): https://www.pcengines.ch/schema/apu4b.pdf

Regarding the DMI matching. The situation is quite complicated since we had wrong SMBIOS naming scheme for some time and we have just fixed it few months ago. Your code still contains very old naming containing capital letters. We do not release such binaries.
Currently DMI product names (depending on BIOS version) found on PC Engines platforms are:

- PC Engines apu1
- apu1
- PC Engines apu2
- apu2
- PC Engines apu3
- apu3
- PC Engines apu4
- apu4

@feckert the ones with shorter names (apu1, apu2 etc.) are official SMBIOS entries that are correct and found on newest firmwares. They won't ever change in the future. The ones with capital letters: APU, APU2, APU3 IMHO should be removed. @pietrushnic please share Your opinion.

**Comment by feckert** [comment]
@miczyg1 could we discuss this in https://github.com/openwrt/openwrt/pull/1232

> Do You mean the S1 button on the back side of board? 

On my APU3 I have a small SMD-Button near the two SIM-Slots. If i press this the system reboots. Maybe you are missing some kmod packages -> **kmod-input-gpio-keys-polled** ?

>Your code still contains very old naming containing capital letters. We do not release such binaries.

Yes i looked at the the driver leds-apu from upstream and copied this dmi informations for APU2 and APU3
 https://git.kernel.org/pub/scm/linux/kernel/git/j.anaszewski/linux-leds.git/tree/drivers/leds/leds-apu.c?h=for-next&id=8cb21086cc6359c6d4f003a788050308afb211ad

**Comment by miczyg1** [comment]
@feckert I did not compile mentioned module. That may be the case. Thanks.

OK, let's move the discussion.

**Pull request closed without merging**